File: javascript/resistor-color-trio/resistor-color-trio.js
const COLOR_VALUES = {
  black: 0,
  brown: 1,
  red: 2,
  orange: 3,
  yellow: 4,
  green: 5,
  blue: 6,
  violet: 7,
  grey: 8,
  white: 9,
};

export class ResistorColorTrio {
  constructor(colors) {
    if (colors.length !== 3) {
      throw new Error('Exactly three colors are required.');
    }
    this.colors = colors.map(color => {
      const lowerCaseColor = color.toLowerCase();
      if (COLOR_VALUES[lowerCaseColor] === undefined) {
        throw new Error(`Invalid color: ${color}`);
      }
      return lowerCaseColor;
    });
  }

  label() {
    const [color1, color2, color3] = this.colors;
    const value1 = COLOR_VALUES[color1];
    const value2 = COLOR_VALUES[color2];
    const exponent = COLOR_VALUES[color3];

    let resistance = (value1 * 10 + value2) * (10 ** exponent);
    let unit = 'ohms';

    if (resistance >= 1000) {
      resistance /= 1000;
      unit = 'kiloohms';
    }

    return `Resistor value: ${resistance} ${unit}`;
  }
}

File: java/resistor-color-trio/src/main/java/ResistorColorTrio.java
class ResistorColorTrio {
    String label(String[] colors) {
        // Map color names to digit values
        java.util.Map<String, Integer> colorMap = java.util.Map.of(
            "black", 0,
            "brown", 1,
            "red", 2,
            "orange", 3,
            "yellow", 4,
            "green", 5,
            "blue", 6,
            "violet", 7,
            "grey", 8,
            "white", 9
        );

        // Extract first three colors
        int firstDigit = colorMap.get(colors[0]);
        int secondDigit = colorMap.get(colors[1]);
        int multiplierPower = colorMap.get(colors[2]);

        // Compute significant value and multiplier
        int significant = firstDigit * 10 + secondDigit;
        long multiplier = (long) Math.pow(10, multiplierPower);

        long resistance = significant * multiplier;

        String unit;
        double displayValue;

        if (resistance >= 1_000_000_000) {
            displayValue = resistance / 1_000_000_000.0;
            unit = "gigaohms";
        } else if (resistance >= 1_000_000) {
            displayValue = resistance / 1_000_000.0;
            unit = "megaohms";
        } else if (resistance >= 1_000) {
            displayValue = resistance / 1_000.0;
            unit = "kiloohms";
        } else {
            displayValue = resistance;
            unit = "ohms";
        }

        // Remove decimal if whole number
        String valueStr;
        if (displayValue == (long) displayValue) {
            valueStr = String.format("%d", (long) displayValue);
        } else {
            valueStr = String.format("%.0f", displayValue);
        }

        return valueStr + " " + unit;
    }
}
